### ISSUE TITLE
Cannot read property 'maxMessageSize' of null

### ISSUE THREAD

**Issue by kalitine** [webrtcHacks/adapter#810]
- [x] I have provided steps to reproduce (e.g. a link to a [jsfiddle](https://jsfiddle.net/))
- [x] I have provided browser name, version and adapter.js version
- [ ] This issue only happens when adapter.js is used

## Versions affected
**Version 65.0.3325.181 (Official Build) Built on Ubuntu , running on Ubuntu 17.10 (64-bit)**

**adapter.js 6.1.5**

## Description
While trying to send a data over RTCDataChannel the following Exception is thrown: `TypeError: Cannot read property 'maxMessageSize' of null`

The bug is pointing to the following code inside webrtc-adapter:
```javascript
var length = data.length || data.size || data.byteLength;
if (length > pc.sctp.maxMessageSize) {
  throw new DOMException('Message too large (can send a maximum of ' +
           pc.sctp.maxMessageSize + ' bytes)', 'TypeError');
}
```
And actually `pc.sctp` is null. This value is possible in [the sctp documentation](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/sctp).

I tried to reproduce this bug with an equivalent code: https://jsfiddle.net/kalitine/1b68eLdr/14951/.
But it works there (`sctp` is not `null`) and I cannot explain why in my lib `sctp` value is `null` (without webrtc-adapter it works).

## Steps to reproduce
https://jsfiddle.net/kalitine/1b68eLdr/14951/

## Expected results
No exception thrown.

## Actual results
`TypeError: Cannot read property 'maxMessageSize' of null`

**Comment by lgrahl** [comment]
The fiddle works perfectly well for me with Chromium version 65.0.3325.181.

**Comment by fippo** [comment]
what is pc.signalingState and pc.iceConnectionState when that happens?
Also pc.localDescription and pc.remoteDescription

**Comment by fippo** [comment]
I could reproduce one case at least...
```
var pc = new RTCPeerConnection();
var dc = pc.createDataChannel("foo");
dc.send("something");
```
With adapter this throws because of .sctp being null. Without adapter this throws `VM166:1 Uncaught DOMException: Failed to execute 'send' on 'RTCDataChannel': RTCDataChannel.readyState is not 'open'`

**Comment by lgrahl** [comment]
We could change that (but there's not much to gain other than ultimate spec conformance - sending is still not possible at that point, regardless of the error).

**Comment by kalitine** [comment]
I'm getting this error when the RTCDataChannel is open. This monday I'll do more tests and I'll provide more information. In my use case the error is thrown during Karma test running. The code is not the same as in the provided jsfiddle, but is very similar logic.

Anyway it ~worse~*costs* nothing to add an sctp null check isn't it?

**Comment by lgrahl** [comment]
> Anyway it worse nothing to add an sctp null check isn't it?

I don't understand. Can you rephrase?

**Comment by kalitine** [comment]
Sorry, a misspell :). It costs nothing to add an sctp null check isn't it?

**Comment by fippo** [comment]
will be done in #812 but i'd still like to understand how this happens

**Comment by kalitine** [comment]
I've done a more detailed test. I apologize, it was my mistake (without webrtc-adapter it does not work neither. Test without webrtc-adapter was running with different code version which worked). I provide here all details which I hope might help you somehow or anybody else to not come to the same issue as I did.

I've have updated the [jsfiddle](https://jsfiddle.net/kalitine/1b68eLdr/14951/) which reproduces this case completely (where the problem is actually the code and not webrtc-adapter).

In all cases we have:
```
Signaling State:  have-local-offer
Ice Connection State:  new
Local Descritpion: {sdp:"v=0 o=- 5874757212316458427 2 IN IP4 127.0.0.1...
SCTP :  undefined
```

The differences are:

**Version 65.0.3325.181 (Official Build) Built on Ubuntu , running on Ubuntu 17.10 (64-bit)**
```
Remote Descritpion:  {sdp:"", type:""}
DOMException: Failed to execute 'send' on 'RTCDataChannel': Could not send data
```

**Version 65.0.3325.181 (Official Build) Built on Ubuntu , running on Ubuntu 17.10 (64-bit)** + **webrtc-adapter**
```
Remote Descritpion:  {sdp:"", type:""}
DOMException: TypeError: Cannot read property 'maxMessageSize' of null
```

**Firefox 59.0.2 (64-bit), running on Ubuntu 17.10 (64-bit)**
```
Remote Descritpion: null
DOMException: An attempt was made to use an object that is not, or is no longer, usable
```

**Firefox 59.0.2 (64-bit), running on Ubuntu 17.10 (64-bit)** + **webrtc-adapter**
```
Remote Descritpion: null
DOMException: pc.sctp is null
```

**Comment by fippo** [comment]
ah... as I suspected, thanks for digging deeper into this.
This will be fixed in #812 which will pass the native error when send is called on a datachannel which is not open yet.

**Comment by ghost** [comment]
@flippo, 

question about your statement:

> when send is called on a datachannel which is not open yet.

Based on the sequence diagram in [WebRTC Spec](https://www.w3.org/TR/webrtc/images/ladder-2party-simple.svg)

`createDataChannel()` <sup>[7]</sup> is way before `setLocalDescription(offer)` <sup>[12]</sup>.
What would be the possible causes why data channel is not yet opened at this <sup>[12]</sup> time?

**Comment by lgrahl** [comment]
Before the connection has been established, calling `createDataChannel` is more of an intent to establish such a channel.

**Comment by fippo** [comment]
original issue should be fixed, please reopen (or holler) if there is still an issue. Thanks!
